File: heap/973KClosestPointsToOrigin.java
import java.util.*;

public class Solution {
    public int[][] kClosest(int[][] points, int k) {
        int[][] ans = new int[k][2];

        PriorityQueue<int[]> pq = new PriorityQueue<int[]>(k, new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                return (b[0]*b[0] + b[1]*b[1]) - (a[0]*a[0] + a[1]*a[1]);
            }
        });

        for(int i = 0; i < points.length; i++) {
            pq.add(points[i]);
            if(pq.size() > k) pq.poll();
        }

        while(k > 0) {
            k--;
            ans[k] = pq.poll();
        }
        return ans;
    }
}

File: heap/973KClosestPointsToOrigin.py
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        heap = []
        
        for x, y in points:
            dist = -(x ** 2 + y ** 2)
            if k == len(heap):
                heapq.heappushpop(heap, (dist, x, y))  # O(logK) as time complexity depend on the size of the heap
            else:
                heapq.heappush(heap, (dist, x, y))  # O(logK)
        return [(x, y) for dist, x, y in heap]